subprojects/parseq/src/main/java/com/linkedin/parseq/internal/DefaultPlanBasedRateLimiter.java

```diff
@@ -0,0 +1,49 @@
+package com.linkedin.parseq.internal;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+
+/**
+ * Default plan based throttling by doing exact match with plan class name.
+ *
+ * @author Min Chen (mnchen@linkedin.com)
+ */
+public class DefaultPlanBasedRateLimiter implements PlanBasedRateLimiter {
+  private final Map<String, Semaphore> _concurrentPlansPerClass;
+
+  public DefaultPlanBasedRateLimiter(Map<String, Integer> planConcurrencyConfig) {
+    _concurrentPlansPerClass = new ConcurrentHashMap<>(planConcurrencyConfig.size());
+    planConcurrencyConfig.forEach((planClass, concurrency) -> {
+      _concurrentPlansPerClass.put(planClass, new Semaphore(concurrency));
+    });
+  }
+
+  @Override
+  public boolean tryAcquire(String planClass) {
+    return _concurrentPlansPerClass.containsKey(planClass) ?
+        _concurrentPlansPerClass.get(planClass).tryAcquire() : true;
+  }
+
+  @Override
+  public boolean tryAcquire(String planClass, long timeout, TimeUnit unit) throws InterruptedException {
+    return _concurrentPlansPerClass.containsKey(planClass) ?
```

Comment:
When tryAcquire a class which is not contains in the map, we should return true? Do we need to insert that class into the map?

Reply:
Yes, we need to return true, because that means that user does not specify concurrency limit to that plan class. We cannot add it to the map since that map should reflect what user configured for parseq Engine.